Try restoring the site editor animation

talldan opened the pull request:
## What?
Attempts to fix https://github.com/WordPress/gutenberg/issues/51903

## How?
~~Restores how the code looked before (relying on `AnimatePresence` to handle entry and exit animations rather than the `display` property). Unfortunately bringing this back reintroduces routing issues on mobile that were solved by https://github.com/WordPress/gutenberg/pull/51558.~~

Since we're not sure how to fix the routing issues on mobile, we're leaving the sidebar present ini the DOM, but hiding it visually and marking it as `inert` to stop all user interactions on the sidebar.

Also resolves some issues with the editor header animation that seem to have been introduced by the addition of Distraction Free mode.

## Testing Instructions
1. Open the site editor
2. Switch between view and edit mode and back again.
3. There should be an animation of the viewport when switching back to view mode.
4. Switch to Distraction Free
5. Switch between view and edit mode and back again.
6. All animations should look nice :) 

## Screenshots or screencast <!-- if applicable -->
### Before
https://github.com/WordPress/gutenberg/assets/677833/57bedbc8-0b83-4984-80e7-a56ad5fd8886


### After
https://github.com/WordPress/gutenberg/assets/677833/e98cf5b2-3fa9-49e4-9896-76affc2a7406

talldan committed:
Try restoring the site editor animation

SaxonF committed:
fix header animation

talldan committed:
Remove accidental addition of layout prop

talldan committed:
tidy up formatting

talldan commented on packages/edit-site/src/components/layout/index.js:
```diff
@@ -247,21 +256,13 @@ export default function Layout() {
 				<div className="edit-site-layout__content">
 					<AnimatePresence initial={ false }>
-						{
+						{ showSidebar && (
```
@kevin940726 mentioned to me that the issue with completely unmounting the sidebar is that it contains hooks that sync the routing state, and while it's unmounted those hooks aren't running, so some navigation doesn't work as expected. This is the reason it was changed in #51558.

@youknowriad Wondering if you have any suggestions for solving that in a different way?

kevin940726 replied:
Specifically, here's a bug I found while testing on mobile:

https://github.com/WordPress/gutenberg/assets/7753001/4a274290-19c0-49ea-9647-bc964eb4077f

The previous sidebar still exists but hidden if we use the browser's back button to go back from the edit mode. It has a higher `z-index` value too so it intercepts the click event (the add new pattern dropdown shouldn't be opened).

There might be a better and more elegant solution for this though! I didn't have time so I settled with the CSS hack 🤦 .

SaxonF replied:
Also related is the issue found [here](https://github.com/WordPress/gutenberg/pull/50565#issuecomment-1564153060) that actually existed before that PR and is potentially related to AnimatePresence not unmounting properly. The added complication is we now need the sidebar mounted on mobile for routing reasons.

youknowriad replied:
> it contains hooks that sync the routing state

Can you clarify which hooks we're talking about here?

talldan replied:
@kevin940726 Do you know the which specific code might be causing that issue?

kevin940726 replied:
I think it's `useSyncPathWithURL` which is called inside the sidebar. I'm not entirely sure if we need it here though, this whole thing is a bit out of my league.

youknowriad replied:
Can't we move `useSyncPathWithURL` to the layout component instead?

kevin940726 replied:
IIRC, that hook relies on `useNavigator` which relies on the Navigator context, so we probably have to move the whole `NavigatorProvider` too. Last time I tried it failed on a weird error which I can't recall. To minimize the scope of the PR, fixing the CSS seems like an easier solution for now.

SaxonF committed:
fix animate presence issue

SaxonF committed:
Merge branch 'try/restore-site-editor-animation' of https://github.com/WordPress/gutenberg into try/restore-site-editor-animation

jasmussen commented:
GIFs don't do this justice, so here'a s video showing the smooth animation restored:

https://github.com/WordPress/gutenberg/assets/1204802/c8042704-7226-49ff-9aff-9330c6ee3314

This is so important to have in place, so thanks for working on this!

draganescu commented:
> Also resolves some issues with the editor header animation that seem to have been introduced by the addition of Distraction Free mode.

What did exactly distraction free break? @jeryj and I worked a lot on respecting the animation while also adding the hover ones for DFM. It's important to know what we broke in the process :)

draganescu commented:
I tested this and it does fix the animation for the frame and the editor header. However it introduces a weird entry animation when in Distraction Free Mode:

## Trunk

https://github.com/WordPress/gutenberg/assets/107534/77e1183e-053a-4d15-a223-b8a314b599d7

## This PR

https://github.com/WordPress/gutenberg/assets/107534/21a70d04-8e8b-46de-822d-ea904c7a3ece

This PR makes the header in DFM do a weird fade out and the site hub looks out of place with nothing to support it behind. @jeryj you may have some ideas?

jeryj commented on packages/edit-site/src/components/layout/index.js:
```diff
@@ -224,47 +224,45 @@ export default function Layout() {
+									y: '-100%',
+								} }
+								initial={ {
+									opacity: isDistractionFree ? 1 : 0,
```
I don't think we need the `isDistractionFree` check here. It seems like it can be `opacity: 0` for any mode.

jeryj commented on packages/edit-site/src/components/layout/index.js:
```diff
@@ -224,47 +224,45 @@ export default function Layout() {
+								} }
+								initial={ {
+									opacity: isDistractionFree ? 1 : 0,
+									y: '-100%',
```
To bring back the transition from `view` to `isDistractionFree`, we need to change this line from `y: '-100%', to `y: isDistractionFree ? 0 : '-100%',`.

This will set the transition for:
- In view mode && isDistractionFree
- Clicking the canvas
- Transition to isDistractionFree should transition from:
- opacity: 1 -> 0
- transform: y: 0 -> y: -100%

jeryj committed:
Fix animation between sidebar view and distraction free edit view

talldan commented:
> What did exactly distraction free break? @jeryj and I worked a lot on respecting the animation while also adding the hover ones for DFM. It's important to know what we broke in the process :)

@draganescu All I know is that when the regression that was accidentally introduced by the Library was reverted, there was still a remaining issue with the header animation that seemed to be last modified by the distraction free PR.

Worth noting that the library change was merged before distraction free, so it's possible that it masked the problems.

talldan commented:
Status update - I believe this PR still has [the bug being discussed here](https://github.com/WordPress/gutenberg/pull/51956#discussion_r1243187701).

I'm not familiar with the routing code. I'll be going AFK for a while soon and won't have time to look into it, so any help appreciated on this. Please feel free to push to this PR or open up an alternative PR using any code from this PR. 

It might be worth adding an end-to-end test case for the bug to demonstrate the issue (cc @kevin940726).

I think there are really two options for moving forwards:
- Find a solution to the routing issue for mobile.
- Implement the animation of the content element in a different way that leaves the sidebar mounted.

jeryj commented:
> Implement the animation of the content element in a different way that leaves the sidebar mounted.

I can try working on this.

jeryj commented:
When the sidebar exits, should it:
- only fade out
- fade out _and_ slide to the left

The current behavior on this branch is for it to also slide to the left, but I wasn't sure if that was intentional. This screen recording shows both animations. First two clicks are the slide out animation and the second two are only an opacity fadein/out.

https://github.com/WordPress/gutenberg/assets/967608/d0ade692-7e63-45a3-9c3a-599bf5be6fd1

mtias commented:
Only fade, the slide out was a bug previously as well.

jeryj committed:
Leave sidebar present and maintain canvas to
sidebar animation

The sidebar is necessary for routing on mobile so we have to maintain its presence in the DOM. Just hiding it isn't enough though, as it is still able to be reached with keyboard tabs and screen readers. Using the relatively new `inert` property disables the element from user interaction, so we add that when we don't want the sidebar to be shown.

jeryj commented on packages/edit-site/src/components/layout/index.js:
```diff
@@ -224,66 +224,61 @@ export default function Layout() {
 				</motion.div>
 
 				<div className="edit-site-layout__content">
-					<AnimatePresence initial={ false }>
```
[`<AnimatePresence />`](https://www.framer.com/motion/animate-presence/) is only needed when we need to animate removing an element. We're not removing it from the DOM, just hiding it from user interactions, so we don't need to use this anymore.

jeryj commented:
This is ready for review, if anyone wants to hop in :)

kevin940726 commented:
This is what I'm getting on mobile viewport after clicking on a pattern category 😅 

![image](https://github.com/WordPress/gutenberg/assets/7753001/05234187-4781-488a-a67f-d53629f7f7ef)

jeryj commented:
> This is what I'm getting on mobile viewport after clicking on a pattern category 😅

That... could be better 😅 I'll take a look!

jeryj committed:
Fix mobile view for pattern library

On Mobile, the canvas mode wasn't being set to edit when using the pattern library. This updates it to use the showSidbar value instead, keeping it in sync with the inert setting.

jeryj commented:
@kevin940726 Fixed! Thanks for catching that.

draganescu commented:
@jeryj is this PR something that could also address this https://github.com/WordPress/gutenberg/pull/51173#issuecomment-1617635238 ? It seems the site hub is rendered as [Object object]?

draganescu commented:
I don't think this is blocking:

- when distraction free mode is active, when entering edit mode the header does not slide in from the top, but instead it appears suddenly

draganescu reviewed (approved):
I tested this and it works very well for what I tried. I'll leave the final call to @kevin940726 because I am not sure I am looking at all the things in the mobile case.

kevin940726 reviewed (approved):
Tested it and it works! Thanks for picking this up, @jeryj !

tellthemachines commented:
Should this be cherry-picked to 6.3? The issue it fixes is in the 6.3 board.

tellthemachines commented:
I just cherry-picked this PR to the update/second-round-RC1 branch to get it included in the next release: d733cc5785
